CAIOS Reference Artifacts Docs (#1393)
wandb/docs: content/en/guides/core/artifacts/track-external-files.md

---

> description: Track files saved in an external bucket, HTTP file server, or an NFS share.

Reviewer: ```suggestion
description: Track files saved in an external bucket, HTTP file server, or NFS share.
```

weight: 7
---

> Use **reference artifacts** to track and use files saved outside of W&B servers, for example in CoreWeave AI Object Storage, an Amazon Simple Storage Service (Amazon S3) bucket, GCS bucket, Azure blob, HTTP file server, or NFS share.

Reviewer: ```suggestion
Use **reference artifacts** to track and use files saved outside of W&B servers, for example in CoreWeave AI Object Storage, an Amazon Simple Storage Service (Amazon S3) bucket, GCS bucket, Azure blob, HTTP file server, NFS share, or S3-compatible storage like [MinIO](https://github.com/minio/minio) hosted in your cloud or infrastructure on your premises.
```
This is the phrasing that is in the BYOB page, assuming that you can also use reference artifacts with S3-compatible storage? For some reason, we always mention S3-compatible at the very end, rather than right after S3.

> Use **reference artifacts** to track and use files saved outside of W&B servers, for example in CoreWeave AI Object Storage, an Amazon Simple Storage Service (S3) bucket, GCS bucket, Azure blob, HTTP file server, or NFS share.

Reviewer: ```suggestion
Use **reference artifacts** to track and use files stored outside of W&B, such as in CoreWeave AI Object Storage, Amazon S3 storage, GCS storage, Azure blob storage, or in a HTTP file server or an NFS share.
```
Nits: ", for example..." is a comma splice. Also, I don't think you need to spell out S3. CAIOS is also a bucket but it got wordy, so I changed the others to just "storage".

> W&B logs metadata about the externally stored object, such as its ETag, size, and version ID if object versioning is enabled on the bucket. Reference artifact metadata does not leave your system.

Reviewer: ```suggestion
W&B logs metadata about the the object, such as the object's ETag and size. If object versioning is enabled on the bucket, the version ID is also logged.
```

{{% alert %}}

> W&B saves an artifact's files to W&B servers when you log an artifact that does not track external files. This is the default behavior when you log artifacts with the W&B Python SDK.

Reviewer: ```suggestion
If you log an artifact that does not track external files, W&B saves the artifact's files to W&B servers. This is the default behavior when you log artifacts with the W&B Python SDK.
```

> See the [Artifacts start]({{< relref "/guides/core/artifacts/artifacts-walkthrough" >}}) for information on how to save files and directories to W&B servers instead.

Reviewer: ```suggestion
See the [Artifacts quickstart]({{< relref "/guides/core/artifacts/artifacts-walkthrough" >}}) for information on how to save files and directories to W&B servers instead of externally.
```

{{% /alert %}}

The following describes how to construct reference artifacts.

## Track an artifact in an external bucket

> Use the W&B Python SDK to track references to files in your CoreWeave AI Object Storage/S3/GCS/Azure bucket. 

Reviewer: ```suggestion
Use the W&B Python SDK to track references to files stored outside of W&B.
```
Not sure it is worth reiterating all of the supported storage locations (THis also doesn't list NFS or HTTP).

weight: 7
---

> Use **reference artifacts** to track and use files saved outside of W&B servers, for example in CoreWeave AI Object Storage (CAIOS), an Amazon Simple Storage Service (S3) bucket, GCS bucket, Azure blob, HTTP file server, or an NFS share.

Reviewer: Coreweave doesn't use the acronym CAIOS and I suspect we also should not. I doubt we need to explode S3, especially if we are not exploding GCS. I think it is "Azure blob storage container" or "in Azure blob storage".

Reviewer 2: Hmm agreed couldn't find CAIOS in coreweave docs, so most likely its an internal acronym

> W&B logs metadata about the CAIOS/S3/GCS/Azure object such as its ETag, size, and version ID (if object versioning is enabled on the bucket). Reference artifact metadata does not leave your system. 

Reviewer: ```suggestion
W&B logs metadata about the externally stored object, such as its ETag, size, and version ID if object versioning is enabled on the bucket. Reference artifact metadata does not leave your system. 
```
It doesn't leave your system because it is stored in the database, not buckets, right? You could link to https://docs.wandb.ai/guides/hosting/data-security/secure-storage-connector/#data-stored-in-the-central-database-vs-buckets.

Reviewer 2: Yes the metadata is stored on W&B side, not in the bucket itself. The "system" here is a bit misleading, but if that's the convention we use in docs that's fine.

Author: @anandwandb What word or phrase would be more accurate for the second sentence?  "Reference artifact metadata does not leave your system."

{{% alert %}}

> Non reference artifacts save artifacts (such as the model and metadata associated for that model) to W&B Servers, and are the default behavior when you log artifacts with the W&B Python SDK.

Reviewer: "Non reference artifacts" is not ideal. Perhaps  something like "Artifacts other than reference artifacts, such as the model and metadata...., are saved in your instance or team's storage bucket."

> See the [Quick start]({{< relref "/guides/core/artifacts/artifacts-walkthrough" >}}) for information on how to save files and directories to W&B servers instead.

Reviewer: ```suggestion
See the [Artifacts quickstart]({{< relref "/guides/core/artifacts/artifacts-walkthrough" >}}) for information on how to save files and directories to W&B servers instead.
```

{{% /alert %}}

> The following describes how to construct reference artifacts and how to incorporate them into your workflows.

Reviewer: I'm not sure you need this sentence since the next section is right there. Maybe "The following sections show how to and download artifacts" or something like that?

## Track an artifact in an external bucket

Use the W&B Python SDK to track references to files in your CAIOS/S3/GCS/Azure bucket. 

1. Initialize a run with `wandb.init()`.
2. Create an artifact object with `wandb.Artifact()`.
3. Specify the reference to the bucket path with the artifact object's `add_reference()` method.
4. Log the artifact's metadata with `run.log_artifact()`.

```python
import wandb

# Initialize a W&B run
run = wandb.init()

# Create an artifact object
artifact = wandb.Artifact(name="name", type="type")

# Add a reference to the bucket path
artifact.add_reference(uri = "uri/to/your/bucket/path")

# Log the artifact's metadata
run.log_artifact(artifact)
run.finish()
```

> Suppose you have a bucket with the following directory structure in your Amazon S3 bucket:

Reviewer: ```suggestion
Suppose your bucket has the following directory structure:
```

> ```bash

Reviewer: ```suggestion
```text
```

s3://my-bucket

|datasets/
  |---- mnist/
|models/
  |---- cnn/
```

> Within `mnist/` is a collection of images. Track the your images dataset with `wandb.Artifact.add_reference()`:

Reviewer: ```suggestion
The `datasets/mnist/` directory contains a collection of images. Track the directory as a dataset with `wandb.Artifact.add_reference()`:
```

```python
import wandb

run = wandb.init()
artifact = wandb.Artifact(name="mnist", type="dataset")
artifact.add_reference(uri="s3://my-bucket/datasets/mnist")
run.log_artifact(artifact)
run.finish()
```

> The previous code sample creates a reference artifact `mnist:latest`. Within the W&B App, you can look through the contents of the reference artifact using the file browser, [explore the full dependency graph]({{< relref "/guides/core/artifacts/explore-and-traverse-an-artifact-graph" >}}), and scan through the versioned history of your artifact. The W&B App does not render rich media such as images, audio, and so forth because the data itself is not contained within the artifact.

Reviewer: If you put the paragraph above the example, you could say "This code sample"

{{% alert %}}

> W&B Artifacts support any Amazon S3 compatible interface, including MinIO. The scripts described below work as-is with MinIO, when you set the `AWS_S3_ENDPOINT_URL` environment variable to point at your MinIO server.

Reviewer: You could steal the language and link for MiniIO from https://docs.wandb.ai/guides/hosting/data-security/secure-storage-connector/#availability-matrix.

{{% /alert %}}

{{% alert color="secondary" %}}
By default, W&B imposes a 10,000 object limit when adding an object prefix. You can adjust this limit by specifying `max_objects=` when you call `add_reference()`.
{{% /alert %}}

## Download an artifact from an external bucket

> W&B uses the metadata recorded when the artifact was logged to retrieve the files from the underlying bucket when it downloads a reference artifact. If your bucket has object versioning enabled, W&B will retrieve the object version corresponding to the state of the file at the time an artifact was logged. This means that as you evolve the contents of your bucket, you can still point to the exact iteration of your data a given model was trained on since the artifact serves as a snapshot of your bucket at the time of training.

Reviewer: ```suggestion
W&B  retrieves the files from the underlying bucket when it downloads a reference artifact using the metadata recorded when the artifact is logged. If your bucket has object versioning enabled, W&B retrieves the object version that corresponds to the state of the file at the time an artifact was logged. As you evolve the contents of your bucket, you can always point to the exact version of your data a given model was trained on,  because the artifact serves as a snapshot of your bucket during the training run.
```

> The following code sample shows how to download a reference artifact. Note that the the APIs for downloading artifacts are the same for both reference and non-reference artifacts:

Reviewer: ```suggestion
The following code sample shows how to download a reference artifact. The APIs for downloading artifacts are the same for both reference and non-reference artifacts:
```

{{% /alert %}}

### Add and download an external reference example

The following code snippet shows how to upload a dataset to an Amazon S3 bucket, track it with a reference artifact, and how to download the artifact:

```python
import boto3
import wandb

run = wandb.init()

# Training here...

s3_client = boto3.client("s3")
s3_client.upload_file(file_name="my_model.h5", bucket="my-bucket", object_name="models/cnn/my_model.h5")

# Log the model artifact
model_artifact = wandb.Artifact("cnn", type="model")
model_artifact.add_reference("s3://my-bucket/models/cnn/")
run.log_artifact(model_artifact)
```

At a later point, you can download the model artifact. Specify the name of the artifact and its type:

```python
import wandb

run = wandb.init()
artifact = run.use_artifact(artifact_or_name = "cnn", type="model")
datadir = artifact.download()
```

{{% alert %}}
See the following reports for an end-to-end walkthrough on how to track artifacts by reference for GCP or Azure:

> * [Guide to Tracking Artifacts by Reference](https://wandb.ai/stacey/artifacts/reports/Tracking-Artifacts-by-Reference--Vmlldzo1NDMwOTE)

Author: @anandwandb Is there an end to end report or blog that we can refer to?

* [Working with Reference Artifacts in Microsoft Azure](https://wandb.ai/andrea0/azure-2023/reports/Efficiently-Harnessing-Microsoft-Azure-Blob-Storage-with-Weights-Biases--Vmlldzo0NDA2NDgw)
{{% /alert %}}

## Cloud storage credentials

W&B uses the default mechanism to look for credentials based on the cloud provider you use. Read the documentation from your cloud provider to learn more about the credentials used:

| Cloud provider | Credentials Documentation |
| -------------- | ------------------------- |

> | CoreWeave AI Object Storage (CAIOS)| [CAIOS documentation](https://docs.coreweave.com/docs/products/storage/object-storage/how-to/manage-access-keys) |

Author: @anmolmann Is this the best resource to point users to for accessing CAIOS buckets?

Reviewer 2: @ngrayluna My handle is @trane293 :)

Reviewer 2: This is a better page since we only support console tokens right now: https://docs.coreweave.com/docs/products/storage/object-storage/how-to/manage-access-keys/cloud-console-tokens

artifact_dir = artifact.download()
```

{{% alert %}}
W&B recommends that you enable 'Object Versioning' on your storage buckets if you overwrite files as part of your workflow. With versioning enabled on your buckets, artifacts with references to files that have been overwritten will still be intact because the older object versions are retained. 

Based on your use case, read the instructions to enable object versioning: [AWS](https://docs.aws.amazon.com/AmazonS3/latest/userguide/manage-versioning-examples.html), [GCP](https://cloud.google.com/storage/docs/using-object-versioning#set), [Azure](https://learn.microsoft.com/azure/storage/blobs/versioning-enable).
{{% /alert %}}

### Add and download an external reference example

> The following code snippet shows how to upload a dataset to an Amazon S3 bucket, track it with a reference artifact, and how to download the artifact:

Reviewer: ```suggestion
The following code sample uploads a dataset to an Amazon S3 bucket, tracks it with a reference artifact, then downloads itt:
```

```python
import boto3
import wandb

run = wandb.init()

# Training here...

s3_client = boto3.client("s3")
s3_client.upload_file(file_name="my_model.h5", bucket="my-bucket", object_name="models/cnn/my_model.h5")

# Log the model artifact
model_artifact = wandb.Artifact("cnn", type="model")
model_artifact.add_reference("s3://my-bucket/models/cnn/")
run.log_artifact(model_artifact)
```

At a later point, you can download the model artifact. Specify the name of the artifact and its type:

```python
import wandb

run = wandb.init()
artifact = run.use_artifact(artifact_or_name = "cnn", type="model")
datadir = artifact.download()
```

{{% alert %}}

> See the following reports for an end-to-end walkthrough on how to track artifacts by reference for GCP or Azure:

Reviewer: ```suggestion
See the following reports for an end-to-end walkthrough of tracking artifacts by reference for GCP or Azure:
```
If the first link is GCP-only, does that need to be part of the link text?

{{% /alert %}}

## Cloud storage credentials

W&B uses the default mechanism to look for credentials based on the cloud provider you use. Read the documentation from your cloud provider to learn more about the credentials used:

| Cloud provider | Credentials Documentation |
| -------------- | ------------------------- |
| CoreWeave AI Object Storage (CAIOS)| [CAIOS documentation](https://docs.coreweave.com/docs/products/storage/object-storage/how-to/manage-access-keys) |
| AWS            | [Boto3 documentation](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html#configuring-credentials) |
| GCP            | [Google Cloud documentation](https://cloud.google.com/docs/authentication/provide-credentials-adc) |
| Azure          | [Azure documentation](https://learn.microsoft.com/python/api/azure-identity/azure.identity.defaultazurecredential?view=azure-python) |

For AWS, if the bucket is not located in the configured user's default region, you must set the `AWS_REGION` environment variable to match the bucket region.

> {{% alert color="secondary" %}}

Reviewer: These points are somewhat orthogonal to each other as they cover entirely separate policies. What about giving the alert a good title that covers both, and making them two bullets?

Author: I thought about that. The sections are somewhat separated so I think it might get lost.

> Rich media such as images, audio, video, and point clouds may fail to render in the App UI depending on the CORS configuration of your bucket. Allow listing **app.wandb.ai** in your bucket's CORS settings will allow the App UI to properly render such rich media.

Reviewer: ```suggestion
If rich media such as images, audio, video, and point clouds does not render in the App UI, you may need to allowlist `app.wandb.ai` in your bucket's CORS policy.
```

> Panels might fail to render in the App UI for private buckets. If your company has a VPN, you could update your bucket's access policy to whitelist IPs within your VPN.

Reviewer: ```suggestion
If you use private buckets and panels fail to render in the App UI, you may need to allowlist your VPN's IP range within your bucket's network access policy.
```

### Add and download an external reference example

> The following code sample uploads a dataset to an Amazon S3 bucket, tracks it with a reference artifact, then downloads it:

Reviewer: Sometimes we say "Amazon S3" and sometimes we just say "S3". Also, you could be using a MiniIO or other S3-compatible bucket.

Author: The correct product name is Amazon S3. This is what we should be using everywhere.

| Cloud provider | Credentials Documentation |
| -------------- | ------------------------- |

> | CoreWeave AI Object Storage (CoreWeave AI Object Storage)| [CoreWeave AI Object Storage documentation](https://docs.coreweave.com/docs/products/storage/object-storage/how-to/manage-access-keys/cloud-console-tokens) |

Reviewer: ```suggestion
| CoreWeave AI Object Storage| [CoreWeave AI Object Storage documentation](https://docs.coreweave.com/docs/products/storage/object-storage/how-to/manage-access-keys/cloud-console-tokens) |
```

{{% alert color="secondary" %}}
Rich media such as images, audio, video, and point clouds may fail to render in the App UI depending on the CORS configuration of your bucket. Allow listing **app.wandb.ai** in your bucket's CORS settings will allow the App UI to properly render such rich media.

> If rich media such as images, audio, video, and point clouds does not render in the App UI, you may need to allowlist `app.wandb.ai` in your bucket's CORS policy.

Reviewer: ```suggestion
If rich media such as images, audio, video, and point clouds does not render in the App UI, ensure that `app.wandb.ai` is allowlisted in your bucket's CORS policy.
```

artifact.add_reference("file:///mount/datasets/mnist/")
run.log_artifact(artifact)
```
{{% alert color="secondary" %}}

> By default, W&B imposes a 10,000 file limit when adding a reference to a directory. You can adjust this limit by specifying `max_objects=` when you call `add_reference()`.

Reviewer: It would be good to keep this in sync with the previous admonition with the same details.

{{% /alert %}}

> Note the triple slash in the URL. The first component is the `file://` prefix that denotes the use of filesystem references. The second is the path to the dataset, `/mount/datasets/mnist/`.

Reviewer: ```suggestion
Note the triple slash in the URL. The first component is the `file://` prefix that denotes the use of filesystem references. The second begins the path to the dataset, `/mount/datasets/mnist/`.
```

The resulting artifact `mnist:latest` looks and acts like a regular artifact. The only difference is that the artifact only consists of metadata about the files, such as their sizes and MD5 checksums. The files themselves never leave your system.

> You can interact with this artifact just as you would a normal artifact. In the UI, you can browse the contents of the reference artifact using the file browser, explore the full dependency graph, and scan through the versioned history of your artifact. However, the UI will not be able to render rich media such as images, audio, etc. as the data itself is not contained within the artifact.

Reviewer: ```suggestion
You can interact with this artifact just as you would a normal artifact. In the UI, you can browse the contents of the reference artifact using the file browser, explore the full dependency graph, and scan through the versioned history of your artifact. However, the UI cannot render rich media such as images, audio, because the data itself is not contained within the artifact.
```

By default, W&B imposes a 10,000 file limit when adding a reference to a directory. You can adjust this limit by specifying `max_objects=` when you call `add_reference()`.
{{% /alert %}}

> Note the triple slash in the URL. The first component is the `file://` prefix that denotes the use of filesystem references. The second begins the path to the dataset, `/mount/datasets/mnist/`.

Reviewer: ```suggestion
Note the triple slash in the URL. The first component is the `file://` prefix that denotes the use of filesystem references. The third slash begins the path to the dataset, `/mount/datasets/mnist/`.
```

artifact_dir = artifact.download()
```

> For filesystem references, a `download()` operation copies the files from the referenced paths to construct the artifact directory. In the above example, the contents of `/mount/datasets/mnist` is copied into the directory `artifacts/mnist:v0/`. If an artifact contains a reference to a file that was overwritten, then `download()` will throw an error as the artifact can no longer be reconstructed.

Reviewer: Maybe a scrub for "as" when it means "because", because I think the latter is easier to parse. Also,  I think this paragraph would benefit from beginning with singular "For a filesystem reference, a...."

> Putting it all together, you can use the following code to track a dataset under a mounted filesystem that feeds into a training job:

Reviewer: ```suggestion
Putting it all together, this example tracks a dataset in a mounted filesystem and feeds it into a training job:
```
(But only if this is correct and doesn't change the meaning)

artifact_dir = artifact.download()
```

> For a filesystem reference, a `download()` operation copies the files from the referenced paths to construct the artifact directory. In the above example, the contents of `/mount/datasets/mnist` is copied into the directory `artifacts/mnist:v0/`. If an artifact contains a reference to a file that was overwritten, then `download()` will throw an error because the artifact can no longer be reconstructed.

Reviewer: ```suggestion
For a filesystem reference, a `download()` operation copies the files from the referenced paths to construct the artifact directory. In the above example, the contents of `/mount/datasets/mnist` are copied into the directory `artifacts/mnist:v0/`. If an artifact contains a reference to a file that was overwritten, `download()` will throw an error because the artifact can no longer be reconstructed.
```

# Perform training here...
```

> To track models, log the model artifact after the training script writes the model files to the mount point:

Reviewer: ```suggestion
To track a model, log the model artifact after the training script writes the model files to the mount point:
```
